apps/ex_wire/lib/ex_wire/adapter/tcp.ex

```diff
@@ -23,45 +23,69 @@ defmodule ExWire.Adapter.TCP do
     GenServer.start_link(__MODULE__, %{
       is_outbound: true,
       peer: peer,
-      active: false,
       subscribers: subscribers
     })
   end
 
   @doc """
-  Initialize by opening up a `gen_tcp` connection to given host and port.
+  Sends auth message to a peer
+  """
+  def send_auth_message(pid, auth_data) do
```

Comment:
I'm thinking about public methods and genserver callbacks. Maybe we should separate them (public methods before callbacks)? There is also post by Dave Thomas - https://pragdave.me/blog/2017/07/13/decoupling-interface-and-implementation-in-elixir.html. He suggests to separate  public apis and genserver calbackes into different modules. I tried to do it in `ExWire.Kademlia.Server` and `ExWire.Kademlia` modules.

Reply:
I totally agree. Glad you mentioned that. That's what I was trying to do here. From this function all the way to `disconnect/1` on line 71 are public interface for the TCP module. The rest (from `init` down) are callbacks.

I know this function `send_auth_message/2` is only used internally currently, but I think it could be used externally, which is why I put it among the public interface of TCP. If you feel strongly about it, I can move it to the server functions.

I am also not opposed to separate them into two different modules. I tend to do that only if the GenServers are complex (which I do think this one is). I didn't do it because I was hoping that we would keep simplifying this GenServer since most of the logic on this file need not live here. What do you think?

Reply:
After some more thought. I like the idea of extracting this into a `Server` module. It cleans up things a bit. If it turns out that we don't need it later, we can always change it then. 

Updated in cb19982ed69ad0bab9c26774f452d6e872bd36a4